### PR TITLE
add manage ec2 keys state

### PR DESCRIPTION
**Pull request by tsaridas**
- create_key
- import_key
- delete_key

### CODE REVIEW

**Comment by techhat** [comment]
Might I suggest key_present (which supports both creating and importing) and key_absent?

**Comment by jfindlay** [comment]
I think what @techhat is saying is that `key_present` and `key_absent` are idempotent concepts, thus being stateful operations.

**Comment by isbm** [comment]
But `key_[present|absent]` more sounds like a properties, than actions. I would suggest `add_key` and "delete_key`, where `add_key` can import existing or create new one.

**Comment by tsaridas** [comment]
@techhat @jfindlay 
create key - creates a private key and saves it locally 
import key - imports a given public key to amazon
they cannot be used together
```
import-key-pair:
  boto_ec2.import_key:
    - name: mykeypair9
    - public_key: 'ssh-rsa AAAAB3Nz'
```
```
create-key-pair:
  boto_ec2.create_key:
    - name: mykeypair6
    - save_path: '/root'
```

Having the below doesn't make sense :
```
create-key-pair:
  boto_ec2.create_key:
    - name: mykeypair6
    - public_key: 'ssh-rsa AAAAB3Nz'
    - save_path: '/root'
```

**Comment by isbm** [comment]
@tsaridas I would also drop `save_path` to `private_key`. Something like:

```
add-key-pair:
  boto_ec2.add_key:
    - name: mykeypair6
    - public_key: 'ssh-rsa AAAAB3NZ'
    - private_key: 'file:///root/ec2keys/'
```

**Comment by tsaridas** [comment]
@isbm as I said, you cannot have both publickey and private key.
in the private key case you are creating one and saving it locally
in the public key case you are uploading your public key and using your private key

what can be done is  
- check to see if both of them are enabled to give an error
- check to see which is being used and use the correct module function.

but I still think this would cause confusion for someone that does not know what private and public keys are.
Its simple this way.

File: salt/states/boto_ec2.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    '''
+    ret = {'name': name,
+           'result': True,
+           'comment': '',
+           'changes': {}
+           }
+    exists = __salt__['boto_ec2.get_key'](name, region, key, keyid, profile)
+    if not exists:
+        if __opts__['test']:
+            ret['comment'] = 'The key {0} is set to be created.'.format(name)
```
Comment (techhat) [inline salt/states/boto_ec2.py]: I would probably say "imported" instead of "created" here.

**Comment by techhat** [comment]
@tsaridas I would venture to say that if somebody doesn't understand how public and private keys work or are used will have a difficult time managing any non-Windows instances on EC2.

Also, I understand that creating and importing are mutually-exclusive, but they still both end in the same result: ensuring the presence of a key in EC2. The logic to auto-detect based on arguments can be added to the state.

Again, just a suggestion. States should be idempotent, and using terminology such as "present" and "absent" follows that logic.

**Comment by isbm** [comment]
@tsaridas
```
you cannot have both publickey and private key.
in the private key case you are creating one and saving it locally
in the public key case you are uploading your public key and using your private key
```
Hmm. I am confused. This statement is self-contradictory: how I can upload my pubkey and use privkey (2b), if I cannot have them both together at first place? Maybe I am missing something, excuse me. :)

So I thought that this is *pair* of the keys and therefore I supposed to have them both, no? Therefore if I define that pair, then private key here is a main player: if I have no one, the new *pair* is generated, private is saved, public is uploaded and used. That said, having **one** method will ensure that key *pair* is valid.

Also I find having one method like @techhat propose is actually easier. And if this is for states, I would agree with @techhat also on terminology, after all.

**Comment by tsaridas** [comment]
@techhat I'll work on another commit.

**Comment by tsaridas** [comment]
@isbm probably I did not explain it correctly.
So, 
create_key: Amazon creates a private key for you and you save it locally. Amazon keeps the public key in order to be added to the servers so you can login by using the private key.

import_key: Amazon imports your public key and you use an already generated private key . The private key is one that you have generated on your own.

What I'm saying is that if a user add both private_key and public_key in the state it doesn't make sense and hence should generate an error since you either want to create a new private key or upload the public of an existing.

What I'll try to do is merge both functions and have an error if user did a mistake.

**Comment by techhat** [comment]
I like `save_path` better than `private_key`, since `private_key` sounds like you're referencing an existing key. And `save_path` makes sense in the context of creating a new key pair, since we know that we are saving a private key (which can be used to create the corresponding public key).

**Comment by isbm** [comment]
@tsaridas Yes. So following @techhat terminology, would it work if the following would happen:
```
# This saves private key
add-key-pair:
  boto_ec2.key_present:
    - name: mykeypair6
    - private_key: 'file:///root/ec2keys/'

# This uploads the public key
add-key-pair:
  boto_ec2.key_present:
    - name: mykeypair6
    - public_key: 'ssh-rsa AAAAB3NZ'

# This raises an error, because pub/priv together cannot be configured
add-key-pair:
  boto_ec2.key_present:
    - name: mykeypair6
    - public_key: 'ssh-rsa AAAAB3NZ'
    - private_key: 'file:///root/ec2keys/'
```

What do you think? @techhat ?

**Comment by techhat** [comment]
@isbm, other than the `private_key` name, I would agree.

**Comment by isbm** [comment]
@techhat Makes sense. How about `location` as an alternative option to `save_path`?

**Comment by techhat** [comment]
@isbm either is fine with me.

**Comment by isbm** [comment]
@techhat Not being annoying here :) but I think if we call `key_present` then `private_key` will make sure the key is there. And how — is it up to user? It will just create one. And if I have already a private key, why not just referencing it indeed?

**Comment by tsaridas** [comment]
I'll go with :

- upload_public
- save_private

Hope that makes you both happy

**Comment by techhat** [comment]
@tsaridas, sounds good.

File: salt/states/boto_ec2.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    '''
+    Only load if boto is available.
+    '''
+    if 'boto_ec2.get_key' in __salt__:
+        return 'boto_ec2'
+    else:
+        return False
+
+
+def present(name, save_priv=None, upload_pub=None, region=None, key=None,
```
Comment (isbm) [inline salt/states/boto_ec2.py]: So it is `save_priv` or `save_private` (`upload_pub` vs `upload_public`)? I would like to have it less cryptic and better explained. Could you please expand these parameters?

Thanks in advance!

**Comment by tsaridas** [comment]
silly mistake of mine.
Question :
I would also like to be able to import the public key from a file, as you said with salt://file but I don't really know how to do that .
Can you point me to any functions in existing states that do it ?
Thanks

**Comment by techhat** [comment]
Certainly, @tsaridas. If you grep the Salt source for `salt.utils.fopen()`, you will find a number of examples. In particular, `salt/utils/http.py` has several examples of both reading and writing files.

**Comment by tsaridas** [comment]
I added support for salt:// 
but it is with cp.get_file_str
hope thats ok.
Also, I changed the names to key_present & key_absent

**Comment by jfindlay** [comment]
@tsaridas, @techhat, @isbm, are you all fine with the status of this pull request?

**Comment by tsaridas** [comment]
I'm welcome to suggestions if @techhat @isbm have any.
Other then that I'm ok and don't think I have anything else to add.

**Comment by isbm** [comment]
@jfindlay LGTM. At least later we can open a PR, in case something special is up.

File: salt/states/boto_ec2.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
+# -*- coding: utf-8 -*-
+'''
+Manage EC2
+
+.. versionadded:: TBD
```
Comment (iggy) [inline salt/states/boto_ec2.py]: Boron... unless you're really unlucky

Comment (rallytime) [inline salt/states/boto_ec2.py]: Actually that one should be Beryllium. The current develop branch will be the Beryllium release. :)

Comment (iggy) [inline salt/states/boto_ec2.py]: Bah! I need a little version number cheat sheet stuck on the wall in my cubicle :/ Thanks for clarifying.

Comment (rallytime) [inline salt/states/boto_ec2.py]: Hehe, no worries. We have a giant periodic table poster. :)

**Comment by cachedout** [comment]
Thanks all for the great dialogue and the excellent additions here!

**Pull request merged**